File: mcp-server/src/api-client.ts
/**
 * App-Sink API Client
 * Handles communication with the App-Sink Kubernetes API
 */

import axios, { type AxiosInstance } from 'axios';

export interface Deployment {
  id: string;
  name: string;
  image: string;
  port: number;
  domain?: string;
  replicas: number;
  status: string;
  environment_variables?: Record<string, string>;
  resource_limits?: {
    cpu?: string;
    memory?: string;
  };
  created_at: string;
  updated_at: string;
}

export interface DeployRequest {
  name: string;
  image: string;
  port: number;
  domain?: string;
  replicas?: number;
  environment_variables?: Record<string, string>;
  resource_limits?: {
    cpu?: string;
    memory?: string;
  };
}

export interface AnalyzeRequest {
  git_url?: string;
  local_path?: string;
  branch?: string;
}

export interface AnalyzeResponse {
  language: string;
  framework?: string;
  entry_point?: string;
  port?: number;
  dependencies: string[];
  recommended_resources: {
    cpu: string;
    memory: string;
  };
}

export class AppSinkClient {
  private client: AxiosInstance;

  constructor(baseURL: string, apiKey: string) {
    this.client = axios.create({
      baseURL,
      headers: {
        'Authorization': `Bearer ${apiKey}`,
        'Content-Type': 'application/json',
      },
    });
  }

  // Deployments
  async listDeployments(): Promise<Deployment[]> {
    const response = await this.client.get<Deployment[]>('/api/v1/apps');
    return response.data;
  }

  async getDeployment(name: string): Promise<Deployment> {
    const response = await this.client.get<Deployment>(`/api/v1/apps/${name}`);
    return response.data;
  }

  async createDeployment(data: DeployRequest): Promise<Deployment> {
    const response = await this.client.post<Deployment>('/api/v1/apps', data);
    return response.data;
  }

  async scaleDeployment(name: string, replicas: number): Promise<Deployment> {
    const response = await this.client.put<Deployment>(
      `/api/v1/apps/${name}/scale`,
      { replicas }
    );
    return response.data;
  }

  async deleteDeployment(name: string): Promise<void> {
    await this.client.delete(`/api/v1/apps/${name}`);
  }

  async getDeploymentLogs(name: string, tail = 100): Promise<string> {
    const response = await this.client.get<string>(
      `/api/v1/apps/${name}/logs`,
      { params: { tail } }
    );
    return response.data;
  }

  async rollbackDeployment(name: string): Promise<Deployment> {
    const response = await this.client.post<Deployment>(
      `/api/v1/apps/${name}/rollback`
    );
    return response.data;
  }

  // AI Analysis
  async analyzeRepository(data: AnalyzeRequest): Promise<AnalyzeResponse> {
    const response = await this.client.post<AnalyzeResponse>(
      '/api/v1/analyze',
      data
    );
    return response.data;
  }

  async analyzeAndDeploy(data: AnalyzeRequest & { name: string; domain?: string }) {
    const response = await this.client.post(
      '/api/v1/analyze-and-deploy',
      data
    );
    return response.data;
  }
}

File: cli/app_sink_cli/api_client.py
"""
API client for App-Sink
"""

import requests
from typing import Dict, Optional, List
from .config import config


class APIClient:
    """Client for App-Sink API"""

    def __init__(self):
        self.endpoint = config.endpoint
        self.api_key = config.api_key

        if not self.endpoint or not self.api_key:
            raise ValueError("API endpoint and key not configured. Run: app-sink config init")

        self.session = requests.Session()
        self.session.headers.update({
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json"
        })

    def _request(self, method: str, path: str, **kwargs) -> requests.Response:
        """Make API request"""
        url = f"{self.endpoint}{path}"
        response = self.session.request(method, url, **kwargs)

        if response.status_code >= 400:
            try:
                error = response.json().get("error", response.text)
            except:
                error = response.text
            raise Exception(f"API Error ({response.status_code}): {error}")

        return response

    def health(self) -> Dict:
        """Check API health"""
        return self._request("GET", "/api/v1/health").json()

    def create_app(self, deployment: Dict) -> Dict:
        """Create new app"""
        return self._request("POST", "/api/v1/apps", json=deployment).json()

    def list_apps(self) -> List[Dict]:
        """List all apps"""
        data = self._request("GET", "/api/v1/apps").json()
        return data.get("deployments", [])

    def get_app(self, name: str) -> Dict:
        """Get app details"""
        return self._request("GET", f"/api/v1/apps/{name}").json()

    def update_app(self, name: str, update: Dict) -> Dict:
        """Update app"""
        return self._request("PUT", f"/api/v1/apps/{name}", json=update).json()

    def scale_app(self, name: str, replicas: int) -> Dict:
        """Scale app"""
        return self._request("PUT", f"/api/v1/apps/{name}/scale", json={"replicas": replicas}).json()

    def delete_app(self, name: str) -> Dict:
        """Delete app"""
        return self._request("DELETE", f"/api/v1/apps/{name}").json()

    def get_logs(self, name: str, tail: int = 100) -> str:
        """Get app logs"""
        response = self._request("GET", f"/api/v1/apps/{name}/logs?tail={tail}")
        return response.json().get("logs", "")

    def rollback_app(self, name: str) -> Dict:
        """Rollback app"""
        return self._request("POST", f"/api/v1/apps/{name}/rollback").json()

    def analyze_repo(self, repo_path: str) -> Dict:
        """Analyze repository with AI"""
        return self._request("POST", "/api/v1/analyze", json={"repo_path": repo_path}).json()

    def analyze_and_deploy(self, git_url: str, branch: str = "main", domain: Optional[str] = None, env: Optional[Dict] = None) -> Dict:
        """Analyze and deploy in one step"""
        data = {
            "git_url": git_url,
            "branch": branch
        }
        if domain:
            data["domain"] = domain
        if env:
            data["env"] = env

        return self._request("POST", "/api/v1/analyze-and-deploy", json=data).json()
